src/background/background-backup-wallet-salt-handler.ts

```diff
@@ -0,0 +1,107 @@
+}) {
+  try {
+    const response = await fetchPrivate(
+      `${gaiaHubConfig.url_prefix}${gaiaHubConfig.address}/wallet-migration-data.json`
+    );
+    if (!response.ok) return null;
+    const encrypted = await response.text();
+    const configJSON = (await decryptContent(encrypted, {
+      privateKey: configPrivateKey,
+    })) as string;
+    const config: WalletDataMigrationConfig = JSON.parse(configJSON);
+    return config;
+  } catch (error) {
+    logger.error(error);
+    return null;
+  }
+}
+const walletMigrationFilename = 'wallet-migration-data.json';
+const walletMigrationVersion = '1.0';
+
+async function storeToGaia(configPrivateKey: string, gaiaHubConfig: GaiaHubConfig, salt: string) {
+  const migrationData: WalletDataMigrationConfig = {
+    version: walletMigrationVersion,
+    oldWalletSalt: salt,
+  };
+  const publicKey = getPublicKeyFromPrivate(configPrivateKey);
+  const encrypted = await encryptContent(JSON.stringify(migrationData), { publicKey });
+  await uploadToGaiaHub(
+    walletMigrationFilename,
+    encrypted,
+    gaiaHubConfig,
+    undefined,
+    undefined,
+    undefined,
+    true
+  );
+  return migrationData;
+}
+const saltUsedInAppKeyInGaia = 'salt-used-in-app-key-in-gaia';
+const walletSaltBackup = 'wallet-salt-backup';
+
+async function backupWalletSaltForGaia(secretKey: string) {
+  // If we already stored in gaia and fetched it, just look up localStorage
+  // We did not do the migration yet, check if previous salt (before this release) is in localStorage
+  const localSalt = localStorage.getItem(walletSaltBackup);
+  const wallet = await generateWallet({ secretKey, password: '' });
+
+  if (localStorage.getItem(saltUsedInAppKeyInGaia)) return;
+  if (!wallet) return;
+  const gaiaHubConfig = await createWalletGaiaConfig({ gaiaHubUrl: gaiaUrl, wallet });
+  if (localSalt) {
+    await storeToGaia(wallet.configPrivateKey, gaiaHubConfig, localSalt);
+    localStorage.setItem(saltUsedInAppKeyInGaia, localSalt);
+    return localSalt;
+  }
+  // Fetch Gaia wallet migration data file
+  const migrationData = await fetchWalletMigrationData({
+    configPrivateKey: wallet.configPrivateKey,
+    gaiaHubConfig,
+  });
+  if (!migrationData) return;
+  // Store salt in localStorage so we don't have to fetch Gaia again and that salt is never updated again
+  localStorage.setItem(saltUsedInAppKeyInGaia, migrationData.oldWalletSalt);
+  return migrationData.oldWalletSalt;
+}
+
+export async function backgroundBackupWalletSaltHandler(
+  message: BackgroundActions,
+  sender: chrome.runtime.MessageSender
+) {
+  if (!validateMessagesAreFromExtension(sender)) {
+    logger.error('Error: Received background script msg from ' + sender.url);
+    return;
+  }
+  logger.info(message);
+  switch (message.method) {
+    case InternalMethods.ShareInMemoryKeyToBackground: {
+      const { secretKey } = message.payload;
+      await backupWalletSaltForGaia(secretKey);
+      break;
+    }
```

Comment:
Is there any reason to do this in a separate message handler, rather than adding this to the handler in `backgroundBackupWalletSaltHandler`.

The gist of `backgroundMessageHandler` is that it's a catch all for background msgs (if we have hundreds of events, maybe we'll need to rethink that, though for this small change, I'd just amend the existing `ShareInMemoryKeyToBackground` action)

Reply:
I thought it will be more decoupled this way. And then we will just remove this extra handler when we won't need it (in a couple of months) without touching the "ShareInMemory..." logic. Wdyt ?

Reply:
That makes sense, though now we have to duplicate the source validation.

Up to you.

Reply:
I will change it